src/ann_solo/reader.py

```diff
@@ -259,6 +259,171 @@ def verify_extension(supported_extensions: List[str], filename: str) -> None:
         raise FileNotFoundError(f'File {filename} does not exist')
 
 
+
+
+def read_mzml(source: Union[IO, str]) -> Iterator[MsmsSpectrum]:
+    """
+    Get the MS/MS spectra from the given mzML file.
+
+    Parameters
+    ----------
+    source : Union[IO, str]
+        The mzML source (file name or open file object) from which the spectra
+        are read.
+
+    Returns
+    -------
+    Iterator[MsmsSpectrum]
+        An iterator over the requested spectra in the given file.
+    """
+    with mzml.MzML(source) as f_in:
+        def spectrum_it():
+            for spectrum_dict in f_in:
+                if int(spectrum_dict.get('ms level', -1)) == 2:
+                    yield spectrum_dict
+
+        try:
+            for i, spectrum in enumerate(spectrum_it()):
+                try:
+                    parsed_spectrum = _parse_spectrum_mzml(spectrum)
+                    parsed_spectrum.index = i
+                    parsed_spectrum.is_processed = False
+                    yield parsed_spectrum
+                except ValueError as e:
+                    logger.warning(f'Failed to read spectrum %s: %s',
+                                   spectrum['id'], e)
+        except LxmlError as e:
+            logger.warning('Failed to read file %s: %s', source, e)
+
+
+def _parse_spectrum_mzml(spectrum_dict: Dict) -> MsmsSpectrum:
+    """
+    Parse the Pyteomics spectrum dict.
+
+    Parameters
+    ----------
+    spectrum_dict : Dict
+        The Pyteomics spectrum dict to be parsed.
+
+    Returns
+    -------
+    MsmsSpectrum
+        The parsed spectrum.
+
+    Raises
+    ------
+    ValueError: The spectrum can't be parsed correctly:
+        - Unknown scan number.
+        - Not an MS/MS spectrum.
+        - Unknown precursor charge.
+    """
+
+    spectrum_id = spectrum_dict['id']
+
+    if 'scan=' in spectrum_id:
+        scan_nr = int(spectrum_id[spectrum_id.find('scan=') + len('scan='):])
+    elif 'index=' in spectrum_id:
+        scan_nr = int(spectrum_id[spectrum_id.find('index=') + len('index='):])
+    else:
+        raise ValueError(f'Failed to parse scan/index number')
+
+    if int(spectrum_dict.get('ms level', -1)) != 2:
+        raise ValueError(f'Unsupported MS level {spectrum_dict["ms level"]}')
+
+
+    mz_array = spectrum_dict['m/z array']
+    intensity_array = spectrum_dict['intensity array']
+    retention_time = spectrum_dict['scanList']['scan'][0]['scan start time']
+
+    precursor = spectrum_dict['precursorList']['precursor'][0]
+    precursor_ion = precursor['selectedIonList']['selectedIon'][0]
+    precursor_mz = precursor_ion['selected ion m/z']
+    if 'charge state' in precursor_ion:
+        precursor_charge = int(precursor_ion['charge state'])
+    elif 'possible charge state' in precursor_ion:
+        precursor_charge = int(precursor_ion['possible charge state'])
+    else:
+        raise ValueError('Unknown precursor charge')
```

Comment:
**suggestion:** We don't need to fail on spectra with an unknown precursor charge, and can set it to `None` instead. Then it will be handled [here](https://github.com/bittremieux/ANN-SoLo/blob/master/src/ann_solo/spectral_library.py#L218).